File: apps/web/src/lib/bushcraft.ts
export type BushcraftDiscipline =
  | 'shelter_craft'
  | 'friction_fire'
  | 'cordage_botany'
  | 'woodcraft_carving'
  | 'water_foraging_craft';

export type SkillLevel =
  | 'foundation_beginner'
  | 'intermediate_bushcraft'
  | 'advanced_wilderness';

export type ThermalSafetyStatus =
  | 'adequate_survival_warmth'
  | 'caution_hypothermia_risk'
  | 'hazardous_sub_freezing';

export interface BushcraftProject {
  id: string;
  title: string;
  region: string;
  discipline: BushcraftDiscipline;
  difficulty: SkillLevel;
  estimatedHours: number;
  thermalRatingRValue: number;
  materialsRequired: string[];
  toolRequired: string;
  description: string;
  highlights: string[];
}

export interface ShelterThermalQuery {
  projectId: string;
  ambientTemperatureF: number;
  windSpeedMph: number;
  debrisThicknessInches: number;
  beddingElevationInches: number;
  fireReflectorWall: boolean;
}

export interface ShelterThermalResult {
  projectTitle: string;
  discipline: BushcraftDiscipline;
  effectiveRValue: number;
  estimatedInteriorTempF: number;
  safetyStatus: ThermalSafetyStatus;
  groundConductiveLossWarning: boolean;
  thermalAdvisory: string;
  fieldcraftTips: string[];
}

export interface BushcraftGearItem {
  id: string;
  name: string;
  category: 'cutting' | 'fire' | 'cooking' | 'cordage' | 'shelter' | 'container';
  mandatory: boolean;
  description: string;
}

export const BUSHCRAFT_PROJECTS: BushcraftProject[] = [
  {
    id: 'boreal-debris-hut-shelter',
    title: 'Boreal Forest Debris Hut & Insulated Raised Bed',
    region: 'Northwoods Boreal Forest, Ely, MN',
    discipline: 'shelter_craft',
    difficulty: 'intermediate_bushcraft',
    estimatedHours: 4.5,
    thermalRatingRValue: 8.0,
    materialsRequired: [
      'Deadfall ridgepole',
      'Forked support stakes',
      'Leaf and moss debris (3ft thick)',
      'Pine bough bedding',
    ],
    toolRequired: 'Full-tang carbon steel bushcraft knife & folding saw',
    description:
      'A classic sub-arctic survival shelter built with an A-frame ridgepole, rib struts, and a thick canopy of insulating debris designed to trap radiant body heat.',
    highlights: [
      '3-foot thick insulating leaf layer',
      'Body-contoured thermal coffin interior',
      'Small crawl-in low draft entry hole',
    ],
  },
  {
    id: 'cedar-bow-drill-ember',
    title: 'Northern White Cedar Bow Drill Friction Fire',
    region: 'Adirondack High Peaks, Lake Placid, NY',
    discipline: 'friction_fire',
    difficulty: 'intermediate_bushcraft',
    estimatedHours: 2.0,
    thermalRatingRValue: 0.0,
    materialsRequired: [
      'White cedar spindle & hearthboard',
      'Hardwood bearing block with green leaf lube',
      'Curved branch bow',
      "Tinder fungus or cedar bark bird's nest",
    ],
    toolRequired: 'Carbon steel knife with 90-degree spine',
    description:
      'Mastering primitive kinetic fire friction using resonant northern white cedar timber, carving burn-in notches, and generating cherry-red glowing embers.',
    highlights: [
      'Smoke-to-black-powder friction threshold',
      'Hot glowing punk ember transfer',
      "Gentle blowing bird's nest to flame",
    ],
  },
  {
    id: 'basswood-bast-fiber-cordage',
    title: 'Basswood Inner Bark Two-Ply Reverse Wrap Cordage',
    region: 'Great Smoky Mountains, Gatlinburg, TN',
    discipline: 'cordage_botany',
    difficulty: 'foundation_beginner',
    estimatedHours: 3.0,
    thermalRatingRValue: 0.0,
    materialsRequired: [
      'Soaked basswood inner bark strips',
      'Retting bath bucket or stream pool',
      'Smooth scraping rock or wooden wedge',
    ],
    toolRequired: 'Fixed blade carving knife',
    description:
      'Harvesting spring bast fibers from fallen linden or basswood, retting outer bark layers, and reverse-twist rolling continuous high-tensile wilderness cordage.',
    highlights: [
      'Inner bast peel separation from outer bark',
      'Two-ply reverse twist rolling technique',
      'Splice staggering for continuous rope strength',
    ],
  },
  {
    id: 'mors-kochanski-super-shelter',
    title: 'Mors Kochanski Polar Super Shelter',
    region: 'Allagash Wilderness Waterway, Fort Kent, ME',
    discipline: 'shelter_craft',
    difficulty: 'advanced_wilderness',
    estimatedHours: 6.0,
    thermalRatingRValue: 12.0,
    materialsRequired: [
      "Clear polyethylene painter's plastic",
      'Mylar reflective space blankets',
      'Green pole lean-to framework',
      'Long log reflector fire wall',
    ],
    toolRequired: 'Forest axe, bushcraft knife, and folding bucksaw',
    description:
      'Legendary polar survival engineering utilizing a clear plastic greenhouse front, rear Mylar radiant reflectors, and an exterior raised reflector fire.',
    highlights: [
      'Greenhouse radiant heat trapping',
      'Reflective Mylar wall backer',
      'Sub-zero survival in light clothing',
    ],
  },
  {
    id: 'birch-bark-water-boiling-vessel',
    title: 'Folded Birch Bark Cooking Pot & Stone Boiling',
    region: 'Superior National Forest, Grand Marais, MN',
    discipline: 'water_foraging_craft',
    difficulty: 'advanced_wilderness',
    estimatedHours: 3.5,
    thermalRatingRValue: 0.0,
    materialsRequired: [
      'Flexible paper birch bark sheet',
      'Split spruce root split-lashes',
      'Smooth non-porous granite river cobbles',
      'Split green wood cooking tongs',
    ],
    toolRequired: 'Awl or knife tip and carving blade',
    description:
      'Pliable winter-peeled birch bark origami-folded into a leak-proof basin and boiled safely using superheated fire stones without scorching the bark.',
    highlights: [
      'Origami-folded water-tight seams',
      'Hot stone radiant conduction boiling',
      'Zero scorched bark water physics',
    ],
  },
];

export const BUSHCRAFT_GEAR: BushcraftGearItem[] = [
  {
    id: 'carbon-steel-bushcraft-knife',
    name: 'High-Carbon Steel Full-Tang Bushcraft Knife with Scandi Grind & 90-Degree Spine',
    category: 'cutting',
    mandatory: true,
    description:
      '1095 or O1 carbon steel fixed blade optimized for wood splitting (batonning), precise notch carving, and throwing sparks from a ferro rod.',
  },
  {
    id: 'bushcraft-folding-saw',
    name: 'Aggressive Cross-Cut Folding Saw or Compact Buck Saw',
    category: 'cutting',
    mandatory: true,
    description:
      'Pull-stroke Japanese tooth folding saw for processing ridgepole timbers, framing stakes, and deadwood rounds with minimal calorie expenditure.',
  },
  {
    id: 'ferrocerium-spark-rod',
    name: 'Heavy-Duty 1/2-Inch Ferrocerium Rod with Hardened Steel Striker',
    category: 'fire',
    mandatory: true,
    description:
      'Reliable all-weather ignition tool delivering showers of 5,500°F sparks even after total water submersion or in sub-zero blizzards.',
  },
  {
    id: 'single-wall-stainless-canteen-cup',
    name: '32oz Single-Wall Stainless Steel Bushcraft Canteen with Nesting Cup & Lid',
    category: 'cooking',
    mandatory: true,
    description:
      'Direct-flame safe 304 food-grade stainless steel bottle and nesting cup combination for boiling pathogen-free water and preparing hot infusions.',
  },
  {
    id: 'tarred-marline-bankline',
    name: 'Roll of #36 Braided Tarred Bankline (Cordage & Lashing)',
    category: 'cordage',
    mandatory: true,
    description:
      'Braided nylon cordage treated with petroleum wax and tar, offering 320 lb tensile strength, rot resistance, and exceptional knot-holding hold for lashings.',
  },
  {
    id: 'heavy-canvas-wool-blanket',
    name: 'Heavyweight 100% Virgin Wool Bushcraft Blanket or Oilskin Tarp',
    category: 'shelter',
    mandatory: true,
    description:
      'Dense 800+ GSM natural wool weave that provides convective wind blocking and maintains thermal insulation even when thoroughly damp or near ember sparks.',
  },
];

export function getBushcraftProjects(discipline?: BushcraftDiscipline): BushcraftProject[] {
  if (!discipline) {
    return [...BUSHCRAFT_PROJECTS];
  }
  return BUSHCRAFT_PROJECTS.filter((p) => p.discipline === discipline);
}

export function getBushcraftProjectById(id: string): BushcraftProject | undefined {
  return BUSHCRAFT_PROJECTS.find((p) => p.id === id);
}

export function getBushcraftGear(): BushcraftGearItem[] {
  return [...BUSHCRAFT_GEAR];
}

export function calculateShelterThermal(query: ShelterThermalQuery): ShelterThermalResult {
  const project = BUSHCRAFT_PROJECTS.find((p) => p.id === query.projectId);
  const projectTitle = project ? project.title : 'Custom Fieldcraft Shelter';
  const discipline = project ? project.discipline : 'shelter_craft';

  const baseR = project ? project.thermalRatingRValue : 6.0;
  const debrisR = query.debrisThicknessInches * 0.9;
  const beddingR = query.beddingElevationInches * 0.8;
  const reflectorR = query.fireReflectorWall ? 2.5 : 0;
  const windRPenalty = (query.windSpeedMph / 15) * 0.7;
  const groundRPenalty =
    query.beddingElevationInches < 4 ? (4 - query.beddingElevationInches) * 1.5 : 0;

  const effectiveRValue = Math.max(
    0.5,
    Math.round((baseR + debrisR + beddingR + reflectorR - windRPenalty - groundRPenalty) * 10) / 10
  );

  const bodyLift = Math.min(30, effectiveRValue * 0.85);
  const reflectorTempBoost = query.fireReflectorWall ? 18 : 0;
  const windTempLoss = query.windSpeedMph * 0.3;
  const groundTempLoss =
    query.beddingElevationInches < 4 ? (4 - query.beddingElevationInches) * 4.0 : 0;

  const estimatedInteriorTempF = Math.round(
    query.ambientTemperatureF + bodyLift + reflectorTempBoost - windTempLoss - groundTempLoss
  );

  const groundConductiveLossWarning = query.beddingElevationInches < 4;

  let safetyStatus: ThermalSafetyStatus;
  if (estimatedInteriorTempF >= 50) {
    safetyStatus = 'adequate_survival_warmth';
  } else if (estimatedInteriorTempF >= 32) {
    safetyStatus = 'caution_hypothermia_risk';
  } else {
    safetyStatus = 'hazardous_sub_freezing';
  }

  let thermalAdvisory = '';
  if (safetyStatus === 'adequate_survival_warmth') {
    thermalAdvisory =
      'Shelter envelope provides sufficient thermal retention to maintain core body temperature safely overnight.';
  } else if (safetyStatus === 'caution_hypothermia_risk') {
    thermalAdvisory =
      'Interior temperature approaches hypothermia risk levels. Increase debris thatch density, reduce draft openings, or erect a fire reflector wall.';
  } else {
    thermalAdvisory =
      'Severe hypothermia and frostbite hazard! Interior conditions are sub-freezing. Immediately elevate bedding off the frozen ground and increase thatch wall thickness to at least 24-36 inches.';
  }

  if (groundConductiveLossWarning) {
    thermalAdvisory = `Warning: Critical conductive ground chill detected! Bedding is under 4 inches. ${thermalAdvisory}`;
  }

  const fieldcraftTips = [
    'Thatch debris until you cannot push an arm through without resistance (target 3 feet for sub-freezing bivouacs).',
    'Body-contour the internal sleeping chamber to minimize dead air volume your body has to warm.',
    'Keep your entrance opening minimal and plug with a debris door plug or pack when inside.',
    'Always insulate the floor before building the roof: conduction to the ground drains heat 24x faster than air.',
  ];

  return {
    projectTitle,
    discipline,
    effectiveRValue,
    estimatedInteriorTempF,
    safetyStatus,
    groundConductiveLossWarning,
    thermalAdvisory,
    fieldcraftTips,
  };
}

File: services/chat/src/api/contoso_chat/bushcraft.py
from typing import Any, Optional

from pydantic import BaseModel


class BushcraftProjectModel(BaseModel):
    project_id: str
    title: str
    region: str
    discipline: str
    difficulty: str
    estimated_hours: float
    thermal_rating_r_value: float
    materials_required: list[str]
    tool_required: str
    description: str
    highlights: list[str]


class ShelterThermalRequest(BaseModel):
    project_id: str
    ambient_temperature_f: float = 30.0
    wind_speed_mph: float = 15.0
    debris_thickness_inches: float = 18.0
    bedding_elevation_inches: float = 6.0
    fire_reflector_wall: bool = False


class ShelterThermalResponse(BaseModel):
    project_id: str
    project_title: str
    discipline: str
    effective_r_value: float
    estimated_interior_temp_f: float
    safety_status: str
    ground_conductive_loss_warning: bool
    thermal_advisory: str
    fieldcraft_tips: list[str]


class BushcraftGearRequirement(BaseModel):
    item_id: str
    name: str
    category: str
    mandatory: bool
    purpose: str


class BushcraftIntent(BaseModel):
    action: str  # "projects_list", "project_detail", "thermal_calc", "gear_checklist"
    project_id: Optional[str] = None
    discipline: Optional[str] = None
    region: Optional[str] = None


DEFAULT_BUSHCRAFT_PROJECTS: dict[str, BushcraftProjectModel] = {
    "boreal-debris-hut-shelter": BushcraftProjectModel(
        project_id="boreal-debris-hut-shelter",
        title="Boreal Forest Debris Hut & Insulated Raised Bed",
        region="Ely, MN",
        discipline="shelter_craft",
        difficulty="intermediate_bushcraft",
        estimated_hours=4.5,
        thermal_rating_r_value=8.0,
        materials_required=[
            "Stout ridge pole (10-12 ft, forearm thickness)",
            "Rib poles and lattice branch framework",
            "Dry coniferous duff and dry leaf debris (2-3 feet deep)",
            "Raised bough bedding lattice and birch bark waterproofing shingles",
        ],
        tool_required="Bushcraft folding saw and full-tang carbon steel knife",
        description="Traditional A-frame debris hut engineered for sub-freezing thermal retention, featuring a 45-degree angled rib cage packed with dead leaf debris and an elevated spruce/balsam fir bough mattress to prevent conductive heat loss to frozen soil.",
        highlights=[
            "High thermal insulation without open fire (R-8.0 rating)",
            "Ground insulation critical against conductive chill",
            "Compact internal volume designed to trap metabolic body heat",
        ],
    ),
    "cedar-bow-drill-ember": BushcraftProjectModel(
        project_id="cedar-bow-drill-ember",
        title="Northern White Cedar Bow Drill Friction Fire",
        region="Adirondacks, NY",
        discipline="friction_fire",
        difficulty="intermediate_bushcraft",
        estimated_hours=2.0,
        thermal_rating_r_value=0.0,
        materials_required=[
            "Dry dead-standing northern white cedar hearth board",
            "Cedar spindle (thumb-thickness, 8 inches)",
            "Flexible curved hardwood branch bow (arm-length)",
            "Dense hardwood or stone hand bearing block with lubricating leaf",
            "Tarred bankline or paracord bow string",
            "Tinder bundle of dry shredded cedar bast fiber and yellow birch curls",
        ],
        tool_required="Carbon steel bushcraft knife with 90-degree spine",
        description="Primitive friction fire method utilizing northern white cedar for both spindle and hearth board to generate an 800°F coal. Requires a carved 45-degree V-notch to collect charred wood dust until an ember ignites, transferred into a shredded fibrous bird's nest tinder bundle.",
        highlights=[
            "Matching wood pairing (cedar on cedar) optimizes friction and ignition temperature",
            "Efficient mechanical leverage via bow string and smooth bearing block",
            "Generates glowing coal without matches or modern spark rods",
        ],
    ),
    "basswood-bast-fiber-cordage": BushcraftProjectModel(
        project_id="basswood-bast-fiber-cordage",
        title="Basswood Inner Bark Two-Ply Reverse Wrap Cordage",
        region="Great Smoky Mtns, TN",
        discipline="cordage_botany",
        difficulty="foundation_beginner",
        estimated_hours=3.0,
        thermal_rating_r_value=0.0,
        materials_required=[
            "Harvested green or retted American basswood sapling branches",
            "Water bath for fiber retting and pectin breakdown",
            "Natural plant sizing or beeswax protective coating",
        ],
        tool_required="Full-tang bushcraft knife with Scandi grind",
        description="Traditional bast fiber processing using the soft inner bark (phloem) of basswood (linden). Stripped fibers are peeled into uniform ribbons and twisted using the two-ply reverse wrap technique, yielding rot-resistant, high-tensile survival lashing and net cordage.",
        highlights=[
            "Two-ply counter-twist locks fibers under tension without unraveling",
            "Splicing staggered fibers produces continuous cordage of indefinite length",
            "Exceptional tensile strength for shelter lashing, bow strings, and snares",
        ],
    ),
    "mors-kochanski-super-shelter": BushcraftProjectModel(
        project_id="mors-kochanski-super-shelter",
        title="Mors Kochanski Polar Super Shelter",
        region="Allagash Wilderness, ME",
        discipline="shelter_craft",
        difficulty="advanced_wilderness",
        estimated_hours=6.0,
        thermal_rating_r_value=12.0,
        materials_required=[
            "Heavy-duty clear 6-mil polyethylene plastic sheeting",
            "Mylar reflective emergency space blanket",
            "Spruce/fir poles for lean-to frame and raised timber sleeping platform",
            "Long log fire reflector logs (8 ft, green hardwood)",
            "Tarred bankline for structural lashings and ridge supports",
        ],
        tool_required="Full-tang carbon steel knife, folding saw, and forest axe",
        description="The classic Mors Kochanski northern wilderness survival shelter. Combines a radiant long-fire, log reflector wall, transparent plastic greenhouse wall, and a reflective mylar ceiling lining to trap radiant infrared heat, maintaining 70°F+ interior comfort even in sub-zero winter conditions.",
        highlights=[
            "Greenhouse effect captures radiant heat from long-fire through clear plastic",
            "Mylar ceiling reflects infrared radiant energy directly back onto sleeper",
            "Capable of keeping sleeper warm without sleeping bag in sub-zero temps",
        ],
    ),
    "birch-bark-water-boiling-vessel": BushcraftProjectModel(
        project_id="birch-bark-water-boiling-vessel",
        title="Folded Birch Bark Cooking Pot & Stone Boiling",
        region="Superior National Forest, MN",
        discipline="water_foraging_craft",
        difficulty="advanced_wilderness",
        estimated_hours=3.5,
        thermal_rating_r_value=0.0,
        materials_required=[
            "Fallen paper birch bark sheet (winter or spring harvest)",
            "Spruce root split bindings or wooden pinch pegs",
            "Smooth non-porous granite river stones (fist-sized)",
            "Tongs or green wood split sticks for hot stone transport",
        ],
        tool_required="Bushcraft knife and awl",
        description="Traditional Ojibwe and northern woods water purification craft. A seamless folded birch bark basket or pot holds wild water, brought to a rapid rolling boil by transferring fire-heated granite stones into the vessel without burning through the wet bark.",
        highlights=[
            "Direct heat transfer via hot non-explosive granite stones",
            "Bark remains below ignition temperature while water is in contact",
            "Yields sterile drinking water and primitive hot stews without metal cookware",
        ],
    ),
}

DEFAULT_BUSHCRAFT_GEAR: list[BushcraftGearRequirement] = [
    BushcraftGearRequirement(
        item_id="carbon-steel-bushcraft-knife",
        name="High-Carbon Steel Full-Tang Bushcraft Knife with Scandi Grind & 90-Degree Spine",
        category="cutting_tools",
        mandatory=True,
        purpose="Precision wood carving, batoning firewood, feather sticks, and striking sparks from ferrocerium rods with the sharp 90-degree spine.",
    ),
    BushcraftGearRequirement(
        item_id="bushcraft-folding-saw",
        name="Aggressive Cross-Cut Folding Saw or Compact Buck Saw",
        category="cutting_tools",
        mandatory=True,
        purpose="Bucking ridgepoles, bedding timbers, and processing dead-standing firewood efficiently without wasting metabolic calories.",
    ),
    BushcraftGearRequirement(
        item_id="ferrocerium-spark-rod",
        name="Heavy-Duty 1/2-Inch Ferrocerium Rod with Hardened Steel Striker",
        category="fire_craft",
        mandatory=True,
        purpose="All-weather spark ignition showering 5,500°F sparks to ignite natural tinder, charred cloth, and pitchwood in gale winds or driving rain.",
    ),
    BushcraftGearRequirement(
        item_id="single-wall-stainless-canteen-cup",
        name="32oz Single-Wall Stainless Steel Bushcraft Canteen with Nesting Cup & Lid",
        category="water_purification",
        mandatory=True,
        purpose="Direct flame water boiling, chemical pathogen disinfection, hydration carry, and primitive backcountry stew cooking.",
    ),
    BushcraftGearRequirement(
        item_id="tarred-marline-bankline",
        name="Roll of #36 Braided Tarred Bankline (Cordage & Lashing)",
        category="cordage",
        mandatory=True,
        purpose="High-strength rot-resistant lashings for debris hut ridgepoles, tripod pot-hangers, ridge lines, and emergency tool handle repairs.",
    ),
    BushcraftGearRequirement(
        item_id="heavy-canvas-wool-blanket",
        name="Heavyweight 100% Virgin Wool Bushcraft Blanket or Oilskin Tarp",
        category="thermal_shelter",
        mandatory=True,
        purpose="Spark-resistant thermal insulation, radiant heat retention around open campfires, ground barrier, and improvised emergency pack wrap.",
    ),
]


def get_bushcraft_projects(discipline: Optional[str] = None) -> list[BushcraftProjectModel]:
    projects = list(DEFAULT_BUSHCRAFT_PROJECTS.values())
    if discipline:
        disc_norm = discipline.strip().lower()
        projects = [p for p in projects if p.discipline.lower() == disc_norm]
    return projects


def get_bushcraft_project_by_id(project_id: str) -> Optional[BushcraftProjectModel]:
    normalized = project_id.strip().lower()
    return DEFAULT_BUSHCRAFT_PROJECTS.get(normalized)


def get_bushcraft_gear() -> list[BushcraftGearRequirement]:
    return DEFAULT_BUSHCRAFT_GEAR


def calculate_shelter_thermal(request: ShelterThermalRequest) -> ShelterThermalResponse:
    project = get_bushcraft_project_by_id(request.project_id)
    if not project:
        raise ValueError(f"Bushcraft project '{request.project_id}' not found")

    # Debris insulation and effective R-value calculation
    if request.project_id == "boreal-debris-hut-shelter":
        base_r = (request.debris_thickness_inches / 18.0) * 8.0
        effective_r = base_r + (2.5 if request.fire_reflector_wall else 0.0)
    elif request.project_id == "mors-kochanski-super-shelter":
        base_r = 12.0 + max(0.0, (request.debris_thickness_inches - 12.0) * 0.2)
        effective_r = base_r + (4.0 if request.fire_reflector_wall else 0.0)
    else:
        effective_r = max(0.5, request.debris_thickness_inches * 0.35 + (2.0 if request.fire_reflector_wall else 0.0))

    effective_r_value = round(effective_r, 1)

    # Ground conductive loss assessment: minimum 6 inches of compressed bedding needed
    ground_conductive_loss_warning = request.bedding_elevation_inches < 6.0

    # Interior temperature modeling
    wind_penalty = max(0.0, (request.wind_speed_mph - 5.0) * 0.3)
    if request.fire_reflector_wall:
        reflector_boost = 25.0 if request.project_id == "mors-kochanski-super-shelter" else 15.0
    else:
        reflector_boost = 0.0
    ground_loss_penalty = 8.0 if ground_conductive_loss_warning else 0.0

    interior_temp = (
        request.ambient_temperature_f
        + (effective_r_value * 1.8)
        - wind_penalty
        + reflector_boost
        - ground_loss_penalty
    )
    estimated_interior_temp_f = round(interior_temp, 1)

    # Safety status determination
    if estimated_interior_temp_f >= 55.0:
        safety_status = "safe"
        thermal_advisory = (
            "Shelter thermal equilibrium is safe (>=55°F). Trapped metabolic body heat and radiant reflection provide "
            "sustainable thermal protection. Maintain a small apex vent to exhaust respiration humidity."
        )
    elif estimated_interior_temp_f >= 40.0:
        safety_status = "advisory"
        thermal_advisory = (
            "Marginal thermal retention (40°F-54°F). Wool blanket wrapping and dry insulative clothing layers required. "
            "Consider adding 6-12 inches more debris thickness and raising bedding elevation."
        )
    else:
        safety_status = "danger"
        thermal_advisory = (
            "CRITICAL HYPOTHERMIA RISK: Projected interior temperature is below 40°F. Immediate danger of cold injury "
            "without elevated bough bedding, thicker debris walls (24-36 inches), or a radiant fire reflector wall."
        )

    fieldcraft_tips = [
        "Ground Conduction Rule: The frozen earth steals body heat up to 32 times faster than cold air; maintain 6+ inches of compacted bough or leaf bedding.",
        "Mors Kochanski Debris Metric: Pile debris until it is armpit deep (2 to 3 feet minimum) over the entire framework to ensure winter survival R-value.",
        "Reflector Wall Geometry: Position a vertical log or stone reflector wall 6 to 8 feet opposite the shelter opening to bounce infrared heat directly onto your bed.",
        "Apex Ventilation: Keep a fist-sized ventilation port open at the top of the shelter to prevent condensation from freezing on the inner thatch.",
    ]

    return ShelterThermalResponse(
        project_id=project.project_id,
        project_title=project.title,
        discipline=project.discipline,
        effective_r_value=effective_r_value,
        estimated_interior_temp_f=estimated_interior_temp_f,
        safety_status=safety_status,
        ground_conductive_loss_warning=ground_conductive_loss_warning,
        thermal_advisory=thermal_advisory,
        fieldcraft_tips=fieldcraft_tips,
    )


def detect_bushcraft_intent(message: str) -> Optional[BushcraftIntent]:
    q = message.lower()

    # Disambiguation guardrails against other services
    # 1. Trip planning exclusions
    if any(k in q for k in ["trip plan", "itinerary", "packing list", "plan a 3-day", "plan a trip"]):
        return None

    # 2. First aid medical exclusions
    if any(k in q for k in ["tourniquet", "cpr", "arterial bleeding", "medical triage", "medical symptoms", "fracture", "first aid kit"]):
        return None

    # 3. Fire safety / burn bans exclusions
    if any(k in q for k in ["fire ban", "burn restriction", "burn ban", "stage 1 fire", "stage 2 fire", "campfire safety regulations", "campfire restrictions"]):
        return None

    # 4. Foraging exclusions
    if any(k in q for k in ["forag", "chanterelle", "morel", "wild edible", "huckleberr", "miner's lettuce", "stinging nettle", "mushroom identification", "poisonous plant"]):
        return None

    # 5. Customer support exclusions
    if any(k in q for k in ["refund", "order #", "return label", "tracking", "promo code", "shipping"]):
        return None

    # Core Bushcraft Keywords
    bushcraft_keywords = [
        "bushcraft",
        "debris hut",
        "bow drill",
        "hand drill",
        "friction fire",
        "super shelter",
        "mors kochanski",
        "bast cordage",
        "bast fiber",
        "scandi grind",
        "ferro rod",
        "ferrocerium",
        "batoning",
        "birch bark boiling",
        "traditional survival",
        "primitive fire",
        "try stick",
    ]

    if not any(k in q for k in bushcraft_keywords):
        return None

    # Project Identification
    project_id: Optional[str] = None
    if "debris hut" in q or "boreal" in q:
        project_id = "boreal-debris-hut-shelter"
    elif "bow drill" in q or "friction fire" in q:
        project_id = "cedar-bow-drill-ember"
    elif "bast" in q or "cordage" in q or "basswood" in q:
        project_id = "basswood-bast-fiber-cordage"
    elif "super shelter" in q or "mors kochanski" in q:
        project_id = "mors-kochanski-super-shelter"
    elif "birch bark" in q or "stone boiling" in q:
        project_id = "birch-bark-water-boiling-vessel"

    # Discipline Identification
    discipline: Optional[str] = None
    if any(k in q for k in ["shelter", "debris hut", "super shelter"]):
        discipline = "shelter_craft"
    elif any(k in q for k in ["friction fire", "bow drill", "hand drill", "primitive fire"]):
        discipline = "friction_fire"
    elif any(k in q for k in ["cordage", "bast"]):
        discipline = "cordage_botany"
    elif any(k in q for k in ["boiling", "vessel"]):
        discipline = "water_foraging_craft"

    # Action Determination
    if any(k in q for k in ["thermal", "r-value", "r value", "insulation", "conductive", "ground chill", "calc"]):
        action = "thermal_calc"
    elif any(k in q for k in ["gear", "checklist", "ferro rod", "scandi knife", "mandatory"]) and not project_id:
        action = "gear_checklist"
    elif project_id and any(k in q for k in ["detail", "about", "how do i build", "how to make", "how to construct", "explain", "how does", "build a"]):
        action = "project_detail"
    elif project_id:
        action = "project_detail"
    elif any(k in q for k in ["gear", "tools", "checklist"]):
        action = "gear_checklist"
    else:
        action = "projects_list"

    return BushcraftIntent(
        action=action,
        project_id=project_id,
        discipline=discipline,
    )


def build_bushcraft_prompt(intent: BushcraftIntent) -> str:
    lines = ["Wilderness Bushcraft & Traditional Fieldcraft Tooling:"]
    if intent.project_id:
        proj = get_bushcraft_project_by_id(intent.project_id)
        if proj:
            lines.append(
                f"- Selected Project: {proj.title} ({proj.region})\n"
                f"  Discipline: {proj.discipline} | Difficulty: {proj.difficulty} | Est Hours: {proj.estimated_hours}h\n"
                f"  Thermal Rating: R-{proj.thermal_rating_r_value:.1f} | Required Tool: {proj.tool_required}\n"
                f"  Materials: {', '.join(proj.materials_required)}\n"
                f"  Description: {proj.description}\n"
                f"  Key Highlights: {'; '.join(proj.highlights)}"
            )
    elif intent.action == "gear_checklist":
        gear = get_bushcraft_gear()
        lines.append("- Mandatory Wilderness Bushcraft Kit Checklist:")
        for g in gear:
            lines.append(f"  * {g.name} [{g.category}]: {g.purpose}")
    elif intent.action == "thermal_calc":
        lines.extend([
            "- Shelter Thermal Engineering & Physics:",
            "  * Ground conductive chill draws heat away 32x faster than air; minimum 6-8 inches bough elevation required.",
            "  * Debris thickness of 2-3 feet (24-36 inches) achieves R-8 to R-12 for sub-freezing survival.",
            "  * Mors Kochanski super shelters utilize radiant reflection and greenhouse dynamics for sub-zero comfort.",
        ])
    else:
        projects = get_bushcraft_projects(discipline=intent.discipline)
        lines.append(
            f"- Available Bushcraft Projects ({intent.discipline or 'All Disciplines'}): "
            + "; ".join(f"{p.title} [{p.discipline}, R-{p.thermal_rating_r_value:.1f}]" for p in projects)
        )

    lines.extend([
        "- Core Traditional Bushcraft Principles:",
        "  1. 90-Degree Spine Carbon Knife: Keep edge shaving-sharp with Scandi grind; use square spine for ferro rod striking.",
        "  2. Friction Fire Precision: Pair matched wood densities (cedar/cedar), carved 45-degree V-notch, and dense tinder.",
        "  3. Elevated Bedding: Never sleep directly on bare ground; insulate with minimum 6 inches dry coniferous boughs.",
        "  4. Reflector Walls: Build sturdy log walls to redirect radiant infrared heat into open shelters.",
    ])
    return "\n".join(lines)


def format_bushcraft_response(intent: BushcraftIntent) -> dict[str, Any]:
    if intent.action == "thermal_calc":
        proj_id = intent.project_id or "boreal-debris-hut-shelter"
        thermal_res = calculate_shelter_thermal(
            ShelterThermalRequest(
                project_id=proj_id,
                ambient_temperature_f=30.0,
                wind_speed_mph=15.0,
                debris_thickness_inches=18.0,
                bedding_elevation_inches=6.0,
                fire_reflector_wall=False,
            )
        )
        answer = (
            f"Shelter Thermal Analysis for {thermal_res.project_title}: "
            f"Effective R-value is R-{thermal_res.effective_r_value:.1f} with an estimated interior temperature of "
            f"{thermal_res.estimated_interior_temp_f:.1f}°F. Safety status: {thermal_res.safety_status.upper()}. "
            f"{thermal_res.thermal_advisory}"
        )
        return {
            "answer": answer,
            "bushcraft_info": {
                "action": "thermal_calc",
                "project_id": thermal_res.project_id,
                "project_title": thermal_res.project_title,
                "effective_r_value": thermal_res.effective_r_value,
                "estimated_interior_temp_f": thermal_res.estimated_interior_temp_f,
                "safety_status": thermal_res.safety_status,
                "ground_conductive_loss_warning": thermal_res.ground_conductive_loss_warning,
                "thermal_advisory": thermal_res.thermal_advisory,
                "fieldcraft_tips": thermal_res.fieldcraft_tips,
            },
        }

    if intent.action == "project_detail" and intent.project_id:
        proj = get_bushcraft_project_by_id(intent.project_id)
        if proj:
            answer = (
                f"Bushcraft Project: {proj.title} ({proj.region}). "
                f"Discipline: {proj.discipline} | Difficulty: {proj.difficulty} | Estimated time: {proj.estimated_hours}h. "
                f"Thermal Rating: R-{proj.thermal_rating_r_value:.1f}. Required Tool: {proj.tool_required}. "
                f"Description: {proj.description} Highlights: {'; '.join(proj.highlights)}."
            )
            return {
                "answer": answer,
                "bushcraft_info": {
                    "action": "project_detail",
                    "project_id": proj.project_id,
                    "project": proj.model_dump(),
                },
            }

    if intent.action == "gear_checklist":
        gear = get_bushcraft_gear()
        gear_names = ", ".join(g.name for g in gear)
        answer = (
            f"Mandatory Bushcraft Kit Checklist (6 items): {gear_names}. "
            "High-carbon steel knife with Scandi grind, folding saw, and 1/2-inch ferrocerium rod form the foundation of traditional fieldcraft."
        )
        return {
            "answer": answer,
            "bushcraft_info": {
                "action": "gear_checklist",
                "gear": [g.model_dump() for g in gear],
            },
        }

    # Default: "projects_list"
    projects = get_bushcraft_projects(discipline=intent.discipline)
    summary = "; ".join(f"{p.title} ({p.discipline}, {p.region})" for p in projects)
    answer = (
        f"Contoso Wilderness Bushcraft & Primitive Survival Crafts: {summary}. "
        "Mastering friction fire, bast cordage, debris shelters, and stone boiling builds self-reliance in remote backcountry terrain."
    )
    return {
        "answer": answer,
        "bushcraft_info": {
            "action": "projects_list",
            "projects": [p.model_dump() for p in projects],
        },
    }
